File: DailyChallenges/Scala/LeetCode_71.scala
object Solution {
    def simplifyPath(path: String): String = {
        val elems = path.split('/')
        val n = elems.length
        
        def parse(i:Int, l:List[String]): String = {
            if (i == n) "/" + l.reverse.mkString("/")
            else {
                val e = elems(i)
                e match {
                    case "" => parse(i+1, l)
                    case "." => parse(i+1, l)
                    case ".." => parse(i+1, if (l.isEmpty) l else l.tail)
                    case _ => parse(i+1, e :: l)
                }
            }
        }
        
        parse(0, List.empty[String])
    }
}

File: DailyChallenges/CPP/LeetCode_71.cpp
class Solution {
public:
    string simplifyPath(string path) {
        vector<string> s;
        size_t st = 0;
        while (true) {
            auto ed = path.find("/", st);
            string p = path.substr(st,((ed == string::npos) ? path.size() : ed)-st);
            if(!s.empty() && p == "..") s.pop_back();
            else if(p != "" && p != "." && p != "..") s.push_back(p);
            if (ed == string::npos) break;
            st = ed+1;
        }
        if(s.empty()) return "/";

        string ans;
        for (auto& str: s) ans += "/" + str;
        
        return ans;
    }
};
